### PR TITLE
Support per installation tokens directly in `GitHubStatusPush`

### PR DESCRIPTION
**Pull request by MagicRB**
This PR creates a patch for buildbot supporting a `token` argument being a function, which given a `reports` object, gives back which token to use. It caches this so it doesn't have to re-ask for it and call `render` on each request. 

PR is a draft for now as it's build for buildbot 3, I first need to bump my instance and then rebase the patch, we should probably keep both patches for 3 and 4.

### CODE REVIEW

**Comment by Mic92** [comment]
Yes. Stable NixOS has buildbot 3 and it's usually a painpoint to backport. However if our patches grow bigger over the summer, it might be better to vendor buildbot to reduce maintenance effort.

**Comment by zimbatm** [comment]
> However if our patches grow bigger over the summer, it might be better to vendor buildbot to reduce maintenance effort.

And also the buildbot NixOS module. I noticed that reasoning becomes complicated with both level of modules. Since it's an opinionated distribution of buildbot, we don't need all that flexibility.

**Comment by Mic92** [comment]
@mergify rebase

**Comment by Mic92** [comment]
I made some change how the vendoring works:

```
The overlay will break if buildbot in unstable will change it's python
override in nixpkgs.
This will happen in the next release when sqlalchemy will be update.

Also I have at least three installations where I share nixpkgs instances
across multiple nixos configuration for performance reasons,
where I cannot use overlays defined in a NixOS module.
```

File: nix/0001-Support-per-installation-tokens-in-GithubStatusPush-.patch [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+--- a/master/buildbot/reporters/github.py
++++ b/master/buildbot/reporters/github.py
+@@ -70,7 +70,14 @@ class GitHubStatusPush(ReporterBase):
+         generators=None,
+         **kwargs,
+     ):
+-        token = yield self.renderSecrets(token)
++        headers = {}
++        if not callable(token):
++            token = yield self.renderSecrets(token)
```
Comment (Mic92) [inline nix/0001-Support-per-installation-tokens-in-GithubStatusPush-.patch:23]: I think the renderSecrets interface would be an alternative approach to this actually. One can also define an object that would the same thing we are currently doing.

**Comment by MagicRB** [comment]
@Mic92 should I pull the commuts you made into #219 ?

**Comment by Mic92** [comment]
> @Mic92 should I pull the commits you made into #219 ?

I can just merge together with the actual feature.

File: nix/0001-Support-per-installation-tokens-in-GithubStatusPush-.patch (hunk 1 above)
Comment (Mic92) [reply nix/0001-Support-per-installation-tokens-in-GithubStatusPush-.patch:23]: Mhm. But than it's not per installation...

**Comment by Mic92** [comment]
^ will test this approach now. It will make the patch even simpler.

**Comment by MagicRB** [comment]
Huh, did not know that was a thing, good to know, how did you find that?

**Comment by Mic92** [comment]
I knew about the IRenderable interface. Than I was just looking for a good candidate that implemented it.

**Comment by MagicRB** [comment]
Ah okay, that didnt occur to me, good catch

**Comment by MagicRB** [comment]
I hope the rendering isn't slow

**Comment by Mic92** [comment]
> I hope the rendering isn't slow

It's the same function in the end that also renders the githhub status name.

**Comment by Mic92** [comment]
Depending on what solution buildbot accepts upstream, we can switch than our implementation around. But let's go with this.

**Comment by Mic92** [comment]
@mergify queue

**Pull request merged**